homeassistant/components/elkm1/__init__.py

```diff
@@ -312,16 +302,40 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     return unload_ok
 
 
-async def async_wait_for_elk_to_sync(elk, timeout):
-    """Wait until the elk system has finished sync."""
+async def async_wait_for_elk_to_sync(elk, timeout, conf_host):
+    """Wait until the elk has finished sync. Can fail login or timeout."""
+
+    def login_status(succeeded):
+        nonlocal success
+
+        success = succeeded
+        if succeeded:
+            _LOGGER.info("ElkM1 login succeeded.")
+        else:
+            elk.disconnect()
+            _LOGGER.error("ElkM1 login failed; invalid username or password.")
+            event.set()
+
+    def sync_complete():
+        event.set()
+
+    success = True
+    event = asyncio.Event()
+    elk.add_handler("login", login_status)
+    elk.add_handler("sync_complete", sync_complete)
     try:
         with async_timeout.timeout(timeout):
-            await elk.sync_complete()
-            return True
+            await event.wait()
     except asyncio.TimeoutError:
+        _LOGGER.error(
+            "Timed out after %d seconds while trying to sync with ElkM1 at %s",
+            SYNC_TIMEOUT,
+            conf_host,
+        )
         elk.disconnect()
+        raise ConfigEntryNotReady
```

Comment:
```suggestion
        raise ConfigEntryNotReady from error
```

Reply:
Thx!

I'm working on the test fixes now.